File: R/resnet-bottleneck.R
relu <- function(x) {
  pmax(0, x)
}

BottleneckBlock <- setRefClass(
  "BottleneckBlock",
  fields = list(
    in_ch = "numeric",
    bn_ch = "numeric",
    out_ch = "numeric",
    W1 = "matrix",
    W2 = "matrix",
    W3 = "matrix",
    Ws = "matrix"
  ),
  methods = list(
    initialize = function(in_channels, bottleneck_channels, out_channels) {
      in_ch <<- in_channels
      bn_ch <<- bottleneck_channels
      out_ch <<- out_channels
      W1 <<- matrix(rnorm(in_channels * bottleneck_channels, sd = 0.01), nrow = in_channels, ncol = bottleneck_channels)
      W2 <<- matrix(rnorm(bottleneck_channels * bottleneck_channels, sd = 0.01), nrow = bottleneck_channels, ncol = bottleneck_channels)
      W3 <<- matrix(rnorm(bottleneck_channels * out_channels, sd = 0.01), nrow = bottleneck_channels, ncol = out_channels)
      Ws <<- if (in_channels != out_channels) matrix(rnorm(in_channels * out_channels, sd = 0.01), nrow = in_channels, ncol = out_channels) else NULL
    },
    forward = function(x) {
      identity <- x
      out <- relu(x %*% W1)
      out <- relu(out %*% W2)
      out <- out %*% W3
      if (!is.null(Ws)) {
        identity <- identity %*% Ws
      }
      relu(out + identity)
    }
  )
)

File: Julia/resnet-bottleneck.jl
relu(x) = max.(0, x)

mutable struct BottleneckBlock
    in_ch::Int
    bn_ch::Int
    out_ch::Int
    W1
    W2
    W3
    Ws
end

function BottleneckBlock(in_channels::Int, bottleneck_channels::Int, out_channels::Int)
    W1 = randn(in_channels, bottleneck_channels) .* 0.01
    W2 = randn(bottleneck_channels, bottleneck_channels) .* 0.01
    W3 = randn(bottleneck_channels, out_channels) .* 0.01
    Ws = in_channels != out_channels ? randn(in_channels, out_channels) .* 0.01 : nothing
    BottleneckBlock(in_channels, bottleneck_channels, out_channels, W1, W2, W3, Ws)
end

function forward(block::BottleneckBlock, x)
    identity = x
    out = relu(x * block.W1)
    out = relu(out * block.W2)
    out = out * block.W3
    if block.Ws !== nothing
        identity = identity * block.Ws
    end
    relu(out .+ identity)
end
